base/system/winlogon/rpcserver.c

```diff
@@ -499,20 +499,102 @@ BaseInitiateSystemShutdownEx(
     BOOLEAN bRebootAfterShutdown,
     ULONG dwReason)
 {
+    HANDLE hToken;
+    PTOKEN_PRIVILEGES tpPrivileges;
+    LUID luidSysShutdown;
+    LUID luidRemoteSysShutdown;
+    LPVOID lpAclBuffer;
+    DWORD dwAclSize = 0;
+    
     TRACE("BaseInitiateSystemShutdownEx()\n");
     TRACE("  Message: %wZ\n", lpMessage);
     TRACE("  Timeout: %lu\n", dwTimeout);
     TRACE("  Force: %d\n", bForceAppsClosed);
     TRACE("  Reboot: %d\n", bRebootAfterShutdown);
     TRACE("  Reason: %lu\n", dwReason);
 
-    //FIXME: Verify that the caller actually has the correct privileges
+    //FIXME: Verify that the caller actually has the correct privileges for remote shutdown
+  // Check if thread token exists
+    if (!OpenThreadToken(GetCurrentThread(), TOKEN_ALL_ACCESS, FALSE, &hToken))
+    {
+        // No thread token, try process token next
+        if (GetLastError() == ERROR_NO_TOKEN)
+        {
+            if (!OpenProcessToken(GetCurrentProcess(), TOKEN_ALL_ACCESS, &hToken))
+                return GetLastError();
+        }
+        else
+            return GetLastError();
+    }
+
+    // Get SeShutdownPrivilege
+    if (!LookupPrivilegeValue(NULL, SE_SHUTDOWN_NAME, &luidSysShutdown))
+    {
+        CloseHandle(hToken);
```

Comment:
CloseHandle can set the last error value, so you may want to save it in a variable before the call.